### PR TITLE
Enable generate hashCode and equals source action

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

Feature #597.

image

Prompt to select fields to include:
image

### CODE REVIEW

**Comment by testforstephen** [comment]
Peer PR on eclipse.jdt.ls side eclipse/eclipse.jdt.ls#940

File: src/sourceAction.ts [hunk 1]
```diff
@@ -52,3 +57,69 @@ export function registerCommands(languageClient: LanguageClient) {
...
+            if (result.foundHashCode) {
+                methods.push('\'int hashCode()\'');
+            }
+
+            if (result.foundEquals) {
+                methods.push('\'boolean equals(Object)\'');
+            }
+
+            const ans = await window.showInformationMessage(`Methods ${methods.join(' and ')} already exist in the Class '${result.type}'. `
+                + 'Do you want to regenerate the implementation?', 'YES', 'NO');
```
Comment (fbricon) [inline src/sourceAction.ts]: 'Regenerate', 'Cancel'

File: src/sourceAction.ts [hunk 2]
```diff
@@ -52,3 +57,69 @@ export function registerCommands(languageClient: LanguageClient) {
...
+            const methods = [];
+            if (result.foundHashCode) {
+                methods.push('\'int hashCode()\'');
+            }
+
+            if (result.foundEquals) {
+                methods.push('\'boolean equals(Object)\'');
+            }
+
+            const ans = await window.showInformationMessage(`Methods ${methods.join(' and ')} already exist in the Class '${result.type}'. `
```
Comment (fbricon) [inline src/sourceAction.ts]: if methods.length == 1 => 'exists' else 'exist'

File: src/sourceAction.ts [hunk 3]
```diff
@@ -52,3 +57,69 @@ export function registerCommands(languageClient: LanguageClient) {
...
+                + 'Do you want to regenerate the implementation?', 'YES', 'NO');
+            if (ans !== 'YES') {
+                return;
+            }
+
+            result.settings.regenerate = true;
+        }
+
+        if (result.settings && result.settings.useJ7HashEquals !== undefined) {
+            const template = await window.showQuickPick([
```
Comment (fbricon) [inline src/sourceAction.ts]: I don't think we need to add a popup here. Turn it into a preference, using default template by default

File: src/protocol.ts [hunk 4]
```diff
@@ -147,4 +147,32 @@ export interface AddOverridableMethodParams {
...
+export interface VariableField {
+    bindingKey: string;
+    name: string;
+    type: string;
+}
+
+export interface CheckHashCodeEqualsResponse {
+    type: string;
+    fields: VariableField[];
+    foundHashCode: boolean;
```
Comment (fbricon) [inline src/protocol.ts]: replace 2 booleans with `existingMethods: string[]`
would contain `hashCode` and `equals`

**Comment by fbricon** [comment]
Here are the new preferences controlling the code generation:

* `java.codeGeneration.hashCodeEquals.useJava7Objects`: Use Objects.hash and Objects.equals when generating the hashCode and equals methods. This setting only applies to Java 7 and higher. Defaults to `false`.
* `java.codeGeneration.hashCodeEquals.useInstanceof`: Use 'instanceof' to compare types when generating the hashCode and equals methods. Defaults to `false`.
* `java.codeGeneration.hashCodeEquals.useBlocks`: Use blocks in 'if' statements when generating the hashCode and equals methods. Defaults to `false`.
* `java.codeGeneration.hashCodeEquals.generateComments`: Generate method comments when generating the hashCode and equals methods. Defaults to `false`.

**Pull request merged**